editor/src/components/canvas/dom-walker.ts

```diff
@@ -748,93 +768,82 @@ function walkElements(
         containerRectLazy,
       ),
     }
+    return result
   }
   if (element instanceof HTMLElement) {
     // Determine the uid of this element if it has one.
-    const uidAttribute = getDOMAttribute(element, UTOPIA_UID_KEY)
-    const parentUIDsAttribute = getDOMAttribute(element, UTOPIA_UID_PARENTS_KEY)
-    const originalUIDAttribute = getDOMAttribute(element, UTOPIA_ORIGINAL_ID_KEY)
-    const originalParentUIDsAttribute = getDOMAttribute(element, UTOPIA_UID_ORIGINAL_PARENTS_KEY)
-    const doNotTraverseAttribute = getDOMAttribute(element, UTOPIA_DO_NOT_TRAVERSE_KEY)
-
-    const traverseChildren: boolean = doNotTraverseAttribute !== 'true'
-
-    // Build the path for this element, substituting an index in if there is no uid attribute.
-    function makeIndexElement(): id {
-      return `index-${index}`
-    }
-    const pathElement = Utils.defaultIfNullLazy<id>(uidAttribute, makeIndexElement)
-    const originalPathElement = Utils.defaultIfNull(uidAttribute, originalUIDAttribute)
-    const parentAttribute = Utils.defaultIfNull(parentUIDsAttribute, originalParentUIDsAttribute)
-
-    // Build the unique path for this element.
-    let uniquePath: TemplatePath = uniqueParentPath
-    if (parentUIDsAttribute != null) {
-      uniquePath = TP.appendToPath(uniquePath, parentUIDsAttribute.split('/'))
-    }
-    uniquePath = TP.appendToPath(uniquePath, pathElement)
+    const uids = getUIDsOnDomELement(element)
+    if (uids != null) {
+      const results = uids.map((uidAttribute) => {
```

Comment:
Worth adding a `FIXME` here to save accidentally merging - right now this will mean we'll walk the DOM multiple times for each element (I know you mentioned knowing this, I'm just adding the comment in case that gets forgotten about amongst the other work in here)

Reply:
This is being worked on by Balazs right now, but will be submitted as a separate PR to unblock this one.